invenio_upgrader/engine.py

```diff
@@ -525,16 +466,68 @@ def order_upgrades(self, upgrades, history=None):
 
         return map(lambda x: upgrades[x], topo_order)
 
+    def _parse_plugin_id(self, plugin_id):
+        """Determine repository from plugin id."""
+        m = re.match("(.+)(_\d{4}_\d{2}_\d{2}_)(.+)", plugin_id)
+        if m:
+            return m.group(1)
+        m = re.match("(.+)(_release_)(.+)", plugin_id)
+        if m:
+            return m.group(1)
+
+        raise RuntimeError("Repository could not be determined from "
+                           "the upgrade identifier: %s." % plugin_id)
+
+
+class UpgradeMeta(type):
+    """Base class for all upgrades."""
+
+    def __new__(mcl, name, bases, nmspc):
+        """New."""
+        cls = super(UpgradeMeta, mcl).__new__(mcl, name, bases, nmspc)
+        if name is not 'UpgradeBase':
+            InvenioUpgrader._upgrades[name] = cls()
+        return cls
+
+
+@add_metaclass(UpgradeMeta)
```

Comment:
Alternatively to using metaclass we can stick with using `entry_points` for explicitly.